### PR TITLE
[inductor] switch AotCodeCompiler to new cpp_builder. (take 3)

### PR DESCRIPTION
**Pull request by henrylhtsang**
Summary: This is basically https://github.com/pytorch/pytorch/pull/131304 together with https://github.com/pytorch/pytorch/pull/132594 and absolute path fix for fbcode.

Test Plan: ci

Differential Revision: D60773405

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by xuhancn — APPROVED** [review]

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -2158,37 +2186,62 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
...
-                f.write(f"// Link cmd\n// {link_cmd}\n")
+                # Temp: add command debug code.
+                if config.is_fbcode():
+                    _temp_validate_new_and_old_command(
+                        link_cmd.split(" "), link_cmd_old.split(" ")
+                    )
+
+                log.debug("aot linkage command: %s", link_cmd)
+                if fbcode_aot_cpu_re:
+                    output_so = os.path.splitext(input_path)[0] + ".so"
```
Comment (henrylhtsang) [inline torch/_inductor/codecache.py:2226]: @hl475 we have to add these lines to make sure it uses absolute paths for internal tests. Do you know a way to clean things up a bit in the future?

**Review by chenyang78 — APPROVED** [review]

**Review by desertfire — APPROVED** [review]
@henrylhtsang , thanks a lot for unblocking this!

**Comment by henrylhtsang** [comment]
@xuhancn please don't merge my PR for me next time... I was trying to land it internally...

**Comment by xuhancn** [comment]
> @xuhancn please don't merge my PR for me next time... I was trying to land it internally...

@henrylhtsang Sorry.

**Pull request closed without merging**